src/import-shadow-hash.c

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "storage.h"
+#include "module.h"
+#include "entry.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <shadow.h>
+
+/**
+ * CreateEntryFromShadowHash:
+ * @username: Username containing the shadow hash.
+ * @sdhash: User's hashed shadow password.
+ */
+gboolean CreateEntryFromShadowHash(const char *username, char *sdhash) {
+  // Create a new Cache Entry.
+  CacheEntry *entry = CacheEntryNew();
+  GBytes* gbytes = g_bytes_new(sdhash, strlen(sdhash));
+  // Set the newly created entry hash using the /etc/shadow hash.
+  CacheEntryHashSet(entry, CACHE_ENTRY_ALGORITHM_CRYPT, gbytes, NULL);
+  g_bytes_unref(gbytes);
+  GError *error = NULL;
+  CacheStorage *storage = CacheStorageNew(DEFAULT_STORAGE_PATH);
+  if (!CacheStoragePutEntry(storage, username, entry, &error))
+      goto done;
+
+  done:
+    if (error) {
+      fprintf(stderr, "Failed to create Cache Entry.\n");
+      return FALSE;
+    } else {
+      fprintf(stdout, "Successfully created Cache Entry.\n");
+      return TRUE;
+    }
+}
+
+
+/**
+ * ReadShadowFile:
+ * @username: Username to read hashed password for.
+ */
+gboolean ReadShadowFile(const gchar *username, const gchar *path) {
+  struct spwd *stmpent = NULL;
+  int error = 0;
+  
+  if (!path) {
+    path = "/etc/shadow";
+  }
+
+  FILE *shadowfile = fopen(path, "r");
+  if (shadowfile == NULL) {
+    error = 1;
+    goto done;
+  }
+
+  // stmpent -- Read next shadow entry from STREAM.
+  stmpent = fgetspent(shadowfile);
+  // Loop through /etc/shadow file
+  while (stmpent) {
+    if (!strcmp(stmpent->sp_namp, username)) {
+      // Create entry from shadow hash if username matches an entry.
+      CreateEntryFromShadowHash(username, stmpent->sp_pwdp);
+      break;
+    }
+    // Get the next shadow entry from STREAM
+    fgetspent(shadowfile);
+  }
+  fclose(shadowfile);
+
+  done:
+    if (error) {
+      fprintf(stderr, "Failed to read shadow file.\n");
+    }
+
+  if (!error) {
+    fprintf(stdout, "Successfully read shadow file.\n");
+    return TRUE;
+  } else {
+    fprintf(stderr, "Shadow read failed.\n");
+    return FALSE;
+  }
+}
+
+
+int main(int argc, char *argv[]) {
```

Comment:
Wrap `int main(...) { ... }` with `#ifndef CACHE_IMPORT_TESTING` just like `escalate_helper*` does so you can test this without having conflicting main functions.

In the part of Makefile.am with your test, you'll add:
`import_test_CFLAGS = -DCACHE_IMPORT_TESTING=1 $(AM_CFLAGS)`